draft-ietf-tls-tls13.md

```diff
@@ -5069,6 +5072,58 @@ lacking a "server_name" extension by terminating the connection with a
 "missing_extension" alert.
 
 
+## Protocol Invariants
+
+This section describes invariants that TLS endpoints and intermediaries MUST
+follow. It also applies to earlier versions, which assumed these rules but did
+not document them.
+
+TLS is designed to be securely and compatibly extensible. Newer clients or
+servers, when communicating with newer peers, should negotiate the most
+preferred common parameters with downgrade protection.  In particular, it is a
+security requirement (see {{security-handshake}}) that intermediaries passing
+traffic between a newer client and newer server without terminating TLS are
+unable to influence the handshake. At the same time, deployments update at
+different rates, so a newer client or server MAY continue to support older
+parameters, which would allow it to interoperate with older endpoints.
+
+For this to work, implementations MUST correctly handle extensible fields:
+
+- A client sending a ClientHello MUST support all parameters advertised in it.
+  Otherwise, the server may fail to interoperate by selecting one of those
+  parameters.
+
+- A server receiving a ClientHello MUST correctly ignore all unrecognized
+  cipher suites, extensions, and other parameters. Otherwise, it may fail to
+  interoperate with newer clients.
+
+- An intermediary which terminates a TLS connection MUST behave as a compliant
+  TLS server (to the original client) and as a compliant TLS client (to the
+  original server). In particular, it MUST generate its own ClientHello
+  containing only parameters it understands, and it MUST generate a fresh
+  ServerHello random value, rather than forwarding the endpoint's value.
+
+  In this scenario, the intermediary is a pair of TLS endpoints for purposes
+  of protocol requirements and security analysis. In most TLS deployments this
+  would require configuring one or both endpoints to trust the intermediary's
+  certificate authority.
```

Comment:
I think this item is a bit misguided in that it lumps CDN's with network devices. I say that not to defend my employer :) but rather because tls-terminating intermediaries has been the problem area.  Maybe that confusion can be cleaned up by a better definition in the first sentence.

Reply:
No, TLS-terminating devices have *not* been the problem area for these purposes. It's the non-terminating devices (which forward the ClientHello but then incorrectly try to parse out the response), and the weird "hybrid" cases where a TLS-terminating device has aspects of a non-terminating one. See my slides from IETF 100.

A TLS-terminator is "easy". You take a TLS server and TLS client and connect them back-toback. Your obligations are thus to be a correct server on one end and a correct client on the other. These obligations hold for both CDNs and MITM boxes. Your server needs to ignore unknown values in the ClientHello. Your client needs to not send ClientHello values it doesn't understand.

Thankfully, I haven't heard of any CDNs that broke since you all just do the obvious thing, while some MITM boxes try to be clever about it (like forwarding the original ClientHello along and then sometimes terminate anyway) and make invalid assumptions.

Reply:
Okay, then I still think you made my point.  Should we even talk about TLS-terminating entities, other than to say it's like two TLS connections?  The problems in this protocol were caused by non-terminating, izzat right?

A TLS terminator has no additional requirements beyond any other TLS terminator.  That's a tautology, but it seems to be what you're saying in your comment.

I still think this item needs more clarity.

Reply:
"the intermediary's certificate authority" is a bit weird, since we don't exactly mean "the certificate authority that issued the intermediary's certificate" and rather "a certificate authority controlled by the intermediate".